crates/query-engine/sql/src/sql/ast.rs

```diff
@@ -299,9 +299,35 @@ pub enum Value {
     Variable(String),
 }
 
-/// Scalar type
+/// Scalar type. This include composite types.
+/// This will always be output as a quoted identifier.
+///
+/// This has a few quirks:
+///
+/// * Array types need to be quoted as `"type name"[]` and _not_ `"type name[]"`. Therefore we
+/// track whether a scalar type is supposed to be an array with the separate field `is_array`.
+///
+/// * Quoting of type name identifiers is only supported for the actual type names recorded in
+/// `pg_type`, and _not_ the SQL standard type names. This means that `character varying` is an
+/// acceptable type name, but `"character varying"` is _not_ (Unless of course you do `CREATE TYPE
+/// "character varying" AS (..)`. Spicy).
+///
 #[derive(Debug, Clone, PartialEq)]
-pub struct ScalarType(pub String);
+pub struct ScalarTypeName {
+    pub type_name: String,
```

Comment:
Yeah, there is some `.clone()`ing and `.to_string()`ing going on because of this, especially for `CompositeTypeName` in the next PR. I guess I just didn't know how to solve this in Rust yet.